K2 testing final

Ankith-Confluent opened the pull request:
This PR refactors the librdkafka test suite to improve cloud compatibility by adding configurable options in `test.conf`. This enables better test portability and maintainability across different Kafka environments.

## Changes

### **1. Configurable Test Framework**
- **`test.skip.idempotent`**: Skip tests requiring idempotent producer (replaces `-I` flag)
- **`test.supported.acks`**: Configure supported acknowledgment values (e.g., `-1,0,1`)
- **`test.sleep.multiplier`**: Configurable sleep multiplier for different environments. Multiplies explicit sleep_for() delays for cluster state propagation (set 0 to skip sleeps), different from test.timeout.multiplier which multiplies API operation timeouts (can't be disabled).
- **`test.timeout.multiplier`**: Unified timeout scaling (existing feature, now consistently used)
- **`test.skip.numbers`**: Skip specific tests by number (comma-separated list)

### **2. Topic Cleanup**
- Many tests were creating topics, but not deleting them.
- Added topic delete utility.

### **3. Replication Factor**
- Changed hardcoded replication factors to use broker defaults (`-1`)

### **4. Preprocessor Directives for Version Compatibility**
- Added preprocessor directives to support librdkafka versions 2.0 through 2.11


## 🔧 Configuration Examples

```ini
test.timeout.multiplier=2.0
test.sleep.multiplier=3.0
test.supported.acks=-1,1
test.skip.idempotent=true
test.skip.numbers=0011,0055,0063,0068,0081,0086

```

k-raina committed:
K2 testing commit

 Conflicts:
	tests/0011-produce_batch.c

k-raina committed:
K2 testing commit

k-raina committed:
K2 testing commit

Ankith-Confluent committed:
K2 Fix ( - 81)

Ankith-Confluent committed:
2.11 cherrypick fix

Ankith-Confluent committed:
K2 Fix

Ankith-Confluent committed:
K2 Fix

Ankith-Confluent committed:
k2 11 and 81 fix

Ankith-Confluent committed:
Cherrypicked fix from 2.8 latest

Ankith-Confluent committed:
k2 Fix 2.8 in latest tests

Ankith-Confluent committed:
K2 fix for 0011 and 0081 for 2.11

Ankith-Confluent committed:
Cherry-pick Fix K2

Ankith-Confluent committed:
small fix

Ankith-Confluent committed:
K2 Fix for 2.7

Ankith-Confluent committed:
small fix

Ankith-Confluent committed:
k2 Fix

Ankith-Confluent committed:
k2 Fix

Ankith-Confluent committed:
K2 Fix

Ankith-Confluent committed:
K2 Fix

Ankith-Confluent committed:
K2 Fix

Ankith-Confluent committed:
K2 Fix

Ankith-Confluent committed:
K2 Fix

Ankith-Confluent committed:
K2 Fix

Ankith-Confluent committed:
81 Fix

Ankith-Confluent committed:
minor fix 11 and 59

Ankith-Confluent committed:
59 clean up

Ankith-Confluent committed:
81 test clean up

Ankith-Confluent committed:
fix for 107 and 113

Ankith-Confluent committed:
Fix critical K2 logic bug in 0081-admin.c - inverted condition for delete.retention.ms

Ankith-Confluent committed:
merged 2.8

Ankith-Confluent committed:
Merge branch 'k2_testing_v2.8.0' into k2_testing_final

Ankith-Confluent committed:
Merge k2_testing_v2.8.0_latest into k2_testing_final

- Merge additional K2 improvements from v2.8.0_latest branch
- Includes fixes for tests 0011, 0050, 0081, 0112
- Resolved compilation conflicts in 0081-admin.c

Ankith-Confluent committed:
Merged 2.7

Ankith-Confluent committed:
Merge branch 'k2_testing_v2.5.3-Latest' into k2_testing_final

Ankith-Confluent committed:
Merged 2.4

Ankith-Confluent committed:
Merge branch 'k2_testing_v2.3.0-Latest' into k2_testing_final

Ankith-Confluent committed:
Merge branch 'k2_testing_v2.2.1-Latest' into k2_testing_final

Ankith-Confluent committed:
Merge branch 'k2_testing_v2.1.1-Latest' into k2_testing_final

Ankith-Confluent committed:
Merge branch 'k2_testing_v2.0.2-Latest' into k2_testing_final

Ankith-Confluent committed:
fix regex tests

Ankith-Confluent committed:
ACKS Fix

Ankith-Confluent committed:
rd_sleep Fix

Ankith-Confluent committed:
rd_sleep second fix

Ankith-Confluent committed:
rd_sleep Fix 3

Ankith-Confluent committed:
rd_sleep Fix 4

Ankith-Confluent committed:
rd_sleep Fix 5

Ankith-Confluent committed:
Removing test_k2_cluster for timeout sceneriaos

Ankith-Confluent committed:
Clean up 113 112 1

Ankith-Confluent committed:
K2  related skip fix by adding a new config

Ankith-Confluent committed:
Fix couple of test , 81 mainly

k-raina reviewed (changes requested):
Please update PR description, since there are large amount of changes. It will help with review

Ankith-Confluent committed:
Delete topics (0001 - 0050)

Ankith-Confluent committed:
SSL skips

Ankith-Confluent committed:
delete topics utility

Ankith-Confluent committed:
delete topics utility

Ankith-Confluent committed:
removed delete

Ankith-Confluent committed:
minro commit

Ankith-Confluent committed:
removed delete

Ankith-Confluent committed:
minor bug

Ankith-Confluent committed:
Github Copilot reviews

Ankith-Confluent committed:
minor spelling mistake

k-raina commented on tests/0001-multiobj.c:
```diff
@@ -91,9 +93,9 @@ int main_0001_multiobj(int argc, char **argv) {
-                        TIMING_ASSERT(&t_full, 0, 999);
-                else
+                if (i > 0) 
+                        TIMING_ASSERT(&t_full, 0, tmout_multip(999));
```
Instead of depending on sleep and guesstimates, can we write a function which waits on topic creation/metadata propagation?

k-raina commented on tests/0009-mock_cluster.c:
```diff
@@ -89,6 +89,7 @@ int main_0009_mock_cluster(int argc, char **argv) {
         test_consumer_poll("CONSUME", c, 0, -1, 0, msgcnt, NULL);
 
         rd_kafka_destroy(c);
+
```
nit

k-raina commented on tests/0008-reqacks.c:
```diff
@@ -175,5 +189,6 @@ int main_0008_reqacks(int argc, char **argv) {
                 rd_kafka_destroy(rk);
         }
 
+
```
nit

k-raina commented on tests/0030-offset_commit.c:
```diff
@@ -539,49 +539,54 @@ static void do_nonexist_commit(void) {
-                       0 /* enable.auto.offset.store */, 0 /* sync */,
-                       1 /* use subscribe */);
 
         rd_free(topic);
```
Do we need remove this?

k-raina commented on tests/0013-null-msgs.c:
```diff
@@ -462,6 +464,7 @@ static void test_produce_consume(void) {
         consume_messages_with_queues(testid, topic, partition_cnt, msgcnt);
         verify_consumed_msg_check();
 
+
```
Nit: Maybe try to remove these additional new lines from other places as well?

k-raina commented on tests/0040-io_event.c:
```diff
@@ -106,6 +106,22 @@ int main_0040_io_event(int argc, char **argv) {
         pfd.revents = 0;
+        
+        /* Handle initial rebalance by polling consumer queue directly */
+        for (int i = 0; i < 3; i++) {
```
Was this needed for K1 also?

Ankith-Confluent replied:
Yes it is required

k-raina commented on tests/0041-fetch_max_bytes.c:
```diff
@@ -93,5 +94,11 @@ int main_0041_fetch_max_bytes(int argc, char **argv) {
         rd_kafka_destroy(rk);
 
+        /* Delete the topic */
+        {
```
Why is this code needed?

k-raina commented on tests/0042-many_topics.c:
```diff
@@ -244,6 +247,14 @@ int main_0042_many_topics(int argc, char **argv) {
+        /* Delete all topics */
+        {
+                rd_kafka_t *del_rk = test_create_handle(RD_KAFKA_PRODUCER, NULL);
+                for (i = 0; i < topic_cnt; i++) {
```
Loop doesn't make sense?

k-raina commented on tests/0045-subscribe_update.c:
```diff
@@ -245,7 +281,10 @@ static void do_test_non_exist_and_partchange(void) {
-        test_kafka_topics("--alter --topic %s --partitions 4", topic_a);
+        test_create_partitions(rk, topic_a, 4);
+        
+        test_sleep(2);
```
This file has multiple `test_sleep` . Do we need to add multiplier here also?

k-raina reviewed (changes requested):
First round of review. Left some comments!

Ankith-Confluent committed:
Removed extra lines and delete topics code

Ankith-Confluent committed:
removed extra lines and fixed formatting

Ankith-Confluent committed:
small fix

k-raina reviewed (changes requested):
Thanks for changes! 
Few comments for rest of files.

Ankith-Confluent commented on tests/0050-subscribe_adds.c:
```diff
@@ -33,6 +33,40 @@
+ * @brief Version-aware partition list printing that avoids leader epoch APIs 
+ *        on older versions
+ */
+static void safe_print_partition_list(
```
Yes will make the changes 👍

Ankith-Confluent commented on tests/0076-produce_retry.c:
```diff
@@ -409,32 +409,46 @@ static void do_test_produce_retry_invalid_msg(rd_kafka_mock_cluster_t *mcluster,
         return 0;
 }
 
+int main_0076_produce_retry_idempotent(int argc, char **argv) {
```
Actually this change was made my Milind I think
Basically the idempotent part was shifted to another function so that it can be skipped through a config

Basically creating two new tests, 76 and 76_idempotent

Ankith-Confluent commented on tests/0081-admin.c:
```diff
@@ -1693,7 +1817,7 @@ do_test_DescribeAcls(rd_kafka_t *rk, rd_kafka_queue_t *useq, int version) {
 
         /* Wait for ACL propagation. */
-        rd_sleep(tmout_multip(2));
+        test_sleep(2);
```
test_sleep is the new function which does the same job, we have a new config test.sleep.multiplier which is used in test_sleep instead of test.timeout.multiplier

k-raina replied:
As discussed
- Add comment for need of`test.sleep.multiplier` in test.conf
- Maybe change test_sleep and base_sleep_ms names

Ankith-Confluent commented on tests/0082-fetch_max_bytes.cpp:
```diff
@@ -78,10 +88,23 @@ static void do_test_fetch_max_bytes(void) {
   Test::conf_set(conf, "max.partition.fetch.bytes", "20000000"); /* ~20MB */
-  Test::conf_set(conf, "fetch.max.bytes", "1000000");            /* ~1MB */
-  Test::conf_set(conf, "receive.message.max.bytes", "1000512");  /* ~1MB+512 */
+  if (test_k2_cluster) {
```
Yes I missed this, will make the changes

Ankith-Confluent committed:
Refactor partition list printing functions to improve version safety. Removed deprecated `safe_print_partition_list` and replaced its usage with `test_print_partition_list_with_errors` and `test_print_partition_list_no_errors` across multiple test files.

Ankith-Confluent committed:
Refactor sleep function calls in tests to use sleep_for for consistency and clarity. Updated all instances of test_sleep to sleep_for across multiple test files, ensuring uniformity in sleep duration handling.

Ankith-Confluent committed:
Remove K2 cluster mode references from tests and simplify fetch configuration. Updated test files to eliminate conditional logic related to K2 clusters, ensuring consistent fetch limits across all tests.

Ankith-Confluent committed:
Update fetch configuration and topic creation in tests for consistency. Adjusted fetch.max.bytes and receive.message.max.bytes settings to improve test reliability. Modified topic creation functions to ensure topics are created with the correct parameters across multiple test cases.

Ankith-Confluent committed:
Enhance test stability by adjusting sleep durations and handling timeout-based waits. Updated tests to accommodate cloud environments with higher latencies, ensuring rebalancing operations complete reliably. Expanded error handling in commit callbacks to include additional scenarios that may occur during rebalancing.

Ankith-Confluent committed:
Improve test robustness by enhancing partition comparison logic and adjusting timeout handling for cloud environments. Added checks for partition existence and refined error handling for GROUP resource configurations, ensuring compatibility with managed Kafka services.

Ankith-Confluent committed:
clang-formatted the changes

Ankith-Confluent committed:
Refactor vector syntax in cooperative rebalance tests for consistency.

Ankith-Confluent committed:
clang-format 18 verison

Ankith-Confluent committed:
small styling changes

Ankith-Confluent committed:
Remove unnecessary topic creation call in rkt_cache test to streamline setup process.

milindl commented on tests/0001-multiobj.c:
```diff
@@ -60,6 +60,8 @@ int main_0001_multiobj(int argc, char **argv) {
                 if (!topic)
                         topic = test_mk_topic_name("0001", 0);
 
+                test_create_topic_if_auto_create_disabled(NULL, topic, -1);
```
Shouldn't this be done only once when we're in the `if (!topic)` block above?

milindl commented on tests/0004-conf.c:
```diff
@@ -511,26 +511,19 @@ int main_0004_conf(int argc, char **argv) {
 #endif
-            "client.dns.lookup",
-            "resolve_canonical_bootstrap_servers_only",
+            /* client.dns.lookup was introduced in librdkafka 2.2.0+ - skip
```
nit: Remove this comment, it's only useful for this review

milindl commented on tests/0008-reqacks.c:
```diff
@@ -96,9 +96,23 @@ int main_0008_reqacks(int argc, char **argv) {
         /* Try different request.required.acks settings (issue #75) */
-        for (reqacks = -1; reqacks <= 1; reqacks++) {
+        /* Test all standard acks values, but skip unsupported ones */
+        int start_acks = -1;
```
declare these with `i` at the top of the block

milindl commented on tests/0011-produce_batch.c:
```diff
@@ -628,11 +650,16 @@ static void test_message_single_partition_record_fail(int variation) {
 
-        const char *confs_set_append[] = {"cleanup.policy", "APPEND",
-                                          "compact"};
+        // Modified for Confluent Cloud compatibility:
```
Also, `Modified for Confluent Cloud compatibility` is for the reviewer of the PR, not for someone who reads the code in the future. It's better to say why you made this change and point to relevant documentation links

milindl commented on tests/0022-consume_batch.c:
```diff
@@ -133,7 +139,6 @@ static void do_test_consume_batch(void) {
                 }
 
                 rd_kafka_topic_destroy(rkts[i]);
-                rd_free(topics[i]);
```
Why was this free removed?

Ankith-Confluent committed:
Adjust sleep duration in fast metadata refresh test to improve timing accuracy.

milindl commented on tests/0033-regex_subscribe.c:
```diff
@@ -304,33 +308,48 @@ static int do_test(const char *assignor) {
-        for (i = 0; i < topic_cnt; i++)
+        for (i = 0; i < topic_cnt; i++) {
+                test_create_topic_if_auto_create_disabled(NULL, topics[i], 1);
+                test_wait_topic_exists(NULL, topics[i], tmout_multip(10000));
```
This should be after produce, in the case where we're relying on autocreate. 
But instead of this why not just create the topics explicitly all the time, would that be possible and easier?

milindl commented on tests/0038-performance.c:
```diff
@@ -59,15 +59,23 @@ int main_0038_performance(int argc, char **argv) {
+                TEST_SKIP("acks=1 not supported by this cluster\n");
+                return 0;
+        }
+        const char *acks_value = "1";
```
Variable declaration at top of block

milindl commented on tests/0040-io_event.c:
```diff
@@ -107,6 +107,27 @@ int main_0040_io_event(int argc, char **argv) {
         pfd.events  = POLLIN;
         pfd.revents = 0;
 
+        /* Handle initial rebalance by polling consumer queue directly */
```
Why is it necessary to use the consumer queue and not the fd (because I see rebalance handling there too)?

Ankith-Confluent replied:
The test was getting stuck because we subscribe at line 93 (which triggers a rebalance event), but we don't set up the fd monitoring until line 104. So it was added to handle that first event (initial rebalance).

milindl replied:
We shouldn't get a rebalance event just by subscribing, unless someone is polling the queue or the rk. There seems to be something going wrong somewhere.

Ankith-Confluent replied:
Oh, that’s interesting.
I did some testing and found that the test passes on AK (without the polling code) but hangs on K2. It might be causing a race condition, since it did pass once on K2 when I reduced the sleep multiplier.
I moved the subscribe code to run after enabling the IO events, and now it passes on both. Is this change valid?

milindl commented on tests/0044-partition_cnt.c:
```diff
@@ -61,17 +61,25 @@ static void test_producer_partition_cnt_change(void) {
+
+        rkt = test_create_topic_object(
+            rk, topic, "message.timeout.ms",
+            tsprintf("%d", tmout_multip(msg_timeout_ms)), NULL);
```
Isn't msg_timeout_ms already inclusive of the multiplier?

Ankith-Confluent committed:
Updated timeout duration

Ankith-Confluent committed:
increased time

Ankith-Confluent committed:
Refactor variable initialization

milindl commented on tests/0045-subscribe_update.c:
```diff
@@ -725,65 +784,84 @@ static void do_test_resubscribe_with_regex() {
         TEST_SAY("Subscribing to %s\n", topic1);
         test_consumer_subscribe(rk, topic1);
+
+        sleep_for(3);
```
Question: In this test, why are we sleeping after subscribe and unsubscribe - there is no propagation step needed for subscribe/unsubscribe, so, where is the extra time requirement coming from? Are the await_assignment and await_revocations timing out? I see that test timeout multiplier is  not used in await_*, it's just 30

milindl commented on tests/0045-subscribe_update.c:
```diff
@@ -392,23 +417,49 @@ static void do_test_topic_remove(void) {
         rd_kafka_topic_partition_list_destroy(topics);
 
+        /* Version-specific wait for assignment */
+        if (rd_kafka_version() >= 0x020100ff) {
```
Question: why is the wait version specific, what changed?

milindl commented on tests/0113-cooperative_rebalance.cpp:
```diff
@@ -907,22 +915,28 @@ static void b_subscribe_with_cb_test(rd_bool_t close_consumer) {
   while (true) {
-    Test::poll_once(c1, 500);
-    Test::poll_once(c2, 500);
+    /* Version-specific poll timeouts for cooperative rebalancing */
```
Please add the reason for doing this in the comment, instead of stating what's done in the code

milindl commented on tests/testshared.h:
```diff
@@ -412,5 +420,30 @@ const char *test_consumer_group_protocol();
+                                             void *opaque);
+
+void test_delete_topic_simple(rd_kafka_t *rk, const char *topic_name);
+#endif
```
```suggestion
#endif // _RDKAFKA_H_
```

milindl reviewed (commented):
Mostly looks OK, left some minor comments.

Ankith-Confluent commented on tests/0033-regex_subscribe.c:
```diff
@@ -304,33 +308,48 @@ static int do_test(const char *assignor) {
+                test_create_topic_if_auto_create_disabled(NULL, topics[i], 1);
+                test_wait_topic_exists(NULL, topics[i], tmout_multip(10000));
+                sleep_for(3);
                 test_produce_msgs_easy(topics[i], testid, RD_KAFKA_PARTITION_UA,
```
we can probably reduce the time from 3 to 1, like how we did in test 42.

Ankith-Confluent committed:
Refactor sleep calls to test_wait_for_metadata_propagation for improved metadata synchronization across tests.

Ankith-Confluent committed:
Update test configuration comment to clarify usage of test_wait_for_metadata_propagation for delay multipliers.

Ankith-Confluent committed:
Merge remote-tracking branch 'origin/master' into k2_testing_final

Ankith-Confluent committed:
Refactor variable initialization and formatting for consistency across test files.

Ankith-Confluent committed:
clang fixes

Ankith-Confluent committed:
Refactor tests to improve clarity and functionality, including adjustments to replica handling, consumer subscription order, and configuration parameters for topic creation.

Ankith-Confluent committed:
style fixes

Ankith-Confluent committed:
Increase wait times for ACL propagation and consumer group rebalancing in tests to ensure stability in cloud environments.

Ankith-Confluent committed:
minor fix

Ankith-Confluent committed:
warnings fixes

Ankith-Confluent committed:
clang format

Ankith-Confluent committed:
Build and clang issues

Ankith-Confluent committed:
Fixes to 146 (not required changes)

Ankith-Confluent committed:
small fix

Ankith-Confluent committed:
spacing related commit

Ankith-Confluent committed:
small spacing changes

Ankith-Confluent committed:
146 formatting changes

Ankith-Confluent committed:
changes in timeouts in run-test

Ankith-Confluent committed:
Fixes in valgrind config

Ankith-Confluent committed:
reverted the change to run-test

Ankith-Confluent committed:
valgrind fix

Ankith-Confluent committed:
Fixes regarding windows warnings and timing for 81

Ankith-Confluent committed:
update the timeouts and waits for 22

Ankith-Confluent committed:
test 81 fixes

Ankith-Confluent committed:
style changes 127

Ankith-Confluent committed:
changes to test 40

Ankith-Confluent committed:
81 flaky fixes

Ankith-Confluent committed:
81 scram flaky fixes

Ankith-Confluent committed:
flaky 111 and 77 fixes

Ankith-Confluent committed:
flaky 81 fixes

Ankith-Confluent committed:
flaky fixes

Ankith-Confluent committed:
replication factor fixes

Ankith-Confluent committed:
GC test fixes

Ankith-Confluent committed:
accept  both type of errors

Ankith-Confluent committed:
Remove test.skip.numbers functionality and related code from tests. Update example configuration to reflect this change.

Ankith-Confluent committed:
Remove idempotent producer test skipping logic from producer latency tests and update example configuration accordingly.

Ankith-Confluent committed:
revert the changes

Ankith-Confluent committed:
debug print version

emasab committed:
wait for topic existence

emasab committed:
Use default reprlication factor in test 0103 test do_test_consumer_producer_txn

emasab committed:
Add toolbox deploy script

Adds packaging/tools/deploy_toolbox_pod.py to automate provisioning a
Kubernetes pod or Docker container, building librdkafka from a given
git ref, and running its test suite against an external cluster.

emasab committed:
Explicitly create test topics when auto-creation is disabled

0026-consume_pause.c, 0030-offset_commit.c, and 0103-transactions.c
(do_test_fenced_txn) assumed broker auto topic-creation, which fails on
clusters where it's disabled (e.g. Confluent Cloud).

emasab committed:
Relax offset/lag assertions in 0061 and 0137 tests for control messages

Depending on the implementation, the commit/abort transaction control
message offset may not be stored and instead correspond to the offset
of the last aborted message, so allow for a small tolerance.

emasab committed:
Only relax 0061 consumer_lag assertion by 1 for transactional tests

The -1 slack for control messages should only apply when running with
transactions enabled; non-transactional runs should match calc_lag exactly.

emasab committed:
Relax committed offset assertion in 0103 control message test

Control message offset may be the last aborted message's offset or the
next one depending on the implementation, so accept a range of 1-2
instead of requiring exactly 2.

emasab committed:
Use sleep multiplier for metadata propagation wait
